component/main.jsonnet

```diff
@@ -25,8 +25,29 @@ local memberlist_secret = kube.Secret(params.speaker.secretname) {
   },
 };
 
+local configmap = kube.ConfigMap(params.configmap_name) {
+  metadata+: {
+    namespace: params.namespace,
+  },
+  data: {
+    config:
+      if std.objectHas(params, 'addresses') then std.manifestYamlDoc({
+        "address-pools": [
+          {
+            "name": "default",
+            "protocol": "layer2",
+            "addresses": params.addresses,
+          },
+        ],
+      })
+      else if std.objectHas(params, 'config') then std.manifestYamlDoc(params.config)
```

Comment:
Using an if/else if chain, makes the use of `addresses` and `config` mutual exclusive. If we keep it that way, I suggest to let `config` have precedence (currently `addresses` has the precedence).

Both `addresses` and `config` do not have default values. So we do not have to put too much thoughts into the issue with merging arrays. I would suggest to get rid of the exclusiveness and support both at the same time (`config` being applied on to the structure based on `addresses`

Ensure the error remains.

Reply:
I do not agree on that topic.

Merging both might result in a wrong config (different address-pool naming, causes duplicate entries) and may cause hidden errors, by magic going on the the component the admin does not know. I'd like to force the admin to define this explicit.

If `config` has precedence, this can lead `addresses` keeps unmanaged in the configuration, also lead to unclear situations. I also like to keep it as is.

Do you agree on my view or do we require a debate?

Reply:
I think in this particular case, we should avoid merging `addresses` and `config`, as @megian correctly points out that merging the two may produce duplicate entries or conflicting configurations.

If we keep the two options mutually exclusive, we need to adjust the code to support "unsetting" `addresses` by setting `addresses` to `null`, see the suggestion above. This pattern will then need to be clearly documented to assist users who want to override a generic `addresses`-based configuration with a more specific `config`-based configuration.

Reply:
Since I don't like to set a variable to null, just to get it running, I've added an check to verify never both variables are set at the same time. I think this does fulfill all the expectations in the code without require a huge documentation.

Reply:
This won't help in the case I've outlined.

Let's assume that in the customer-wide config I set `metallb.addresses`, but for a particular cluster I want to use `metallb.config` instead. With the current implementation there's no way to make `metallb.config` the variant that's used, if `metallb.addresses` was set higher-up in the config.

The only way to handle this case is setting `addresses` to `null` to "deconfigure" the customer-wide setting and making sure this case doesn't break the implementation.

Reply:
`metallb.addresses` are most of the time unique to the cluster. The only reason ever to configure `metallb.addresses` on a global level is having multiple clusters with the same address, network isolated. You need a NAT between all network connections, which is worse by design anyway. If I have something like in example vCloud Director with templates, I will most probably use a higher level load balancer instead of MetalLB on layer2. (If not, the case can still be configured, using `metallb.config` instead of `metallb.addresses`)

I take out admin are configuring that on a global level by mistake. I think we can anyway not prevent all dumb config mistakes. Configuring `metallb.addresses` on a global level will mess up all the clusters.

Your proposal is the most robust solution, while not intuitive to me. So do you think it's worth to support this special case instead of simpler and clearer handling?

Reply:
The point about `addresses` being unique to a cluster most of the time is valid, so I guess we don't need to handle the case where we want to override `addresses`.

In any case, we can always adjust the behavior if we find that this variation does not work for some particular case.

Note that currently setting `addresses` to `null` will produce a configuration that may lead to confusing behavior, but the discussion of whether we want to try to guard against configuration mistakes in Commodore components is out of scope for this PR.